### PR TITLE
[CDAP-2923] Added Support for Workflow Token in Spark programs run through workflows.

### PR DESCRIPTION
**Pull request by bdmogal**
Jira: [CDAP-2923](https://issues.cask.co/browse/CDAP-2923)
Build: http://builds.cask.co/browse/CDAP-DUT2240

Note: Currently, workflow token is available in ``beforeSubmit``, ``onFinish``, ``driver`` and ``executor``. However, you can only update it in ``beforeSubmit``, ``onFinish`` and ``driver``

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/NodeValue.java [hunk 1]
```diff
@@ -16,12 +16,14 @@
...
 
+import java.io.Serializable;
+
 /**
  * Multiple nodes in the Workflow can add the same key to the {@link WorkflowToken}.
  * This class provides a mapping from node name to the {@link Value} which was set for the
  * specific key.
  */
-public final class NodeValue {
+public final class NodeValue implements Serializable {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/NodeValue.java]: Includes the `serialVersionUID` field.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/Value.java [hunk 2]
```diff
@@ -16,10 +16,12 @@
 
 package co.cask.cdap.api.workflow;
 
+import java.io.Serializable;
+
 /**
  * Class representing the value of the key in the {@link WorkflowToken}.
  */
-public class Value {
+public class Value implements Serializable {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/Value.java]: Same as above.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java [hunk 3]
```diff
@@ -72,7 +74,8 @@
   protected AbstractSparkContext(SparkSpecification specification, Id.Program programId, RunId runId,
                                  ClassLoader programClassLoader, long logicalStartTime,
                                  Map<String, String> runtimeArguments, DiscoveryServiceClient discoveryServiceClient,
-                                 MetricsContext metricsContext, LoggingContext loggingContext) {
+                                 MetricsContext metricsContext, LoggingContext loggingContext,
+                                 WorkflowToken workflowToken) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java]: So `workflowToken` is nullable, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ClientSparkContext.java [hunk 4]
```diff
@@ -49,12 +50,12 @@
   public ClientSparkContext(Program program, RunId runId, long logicalStartTime, Map<String, String> runtimeArguments,
                             TransactionContext transactionContext, DatasetFramework datasetFramework,
                             DiscoveryServiceClient discoveryServiceClient,
-                            MetricsCollectionService metricsCollectionService) {
+                            MetricsCollectionService metricsCollectionService, WorkflowToken workflowToken) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ClientSparkContext.java]: Is it nullable?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java [hunk 5]
```diff
@@ -93,11 +94,11 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
                                Transaction transaction, DatasetFramework datasetFramework,
                                DiscoveryServiceClient discoveryServiceClient,
                                MetricsCollectionService metricsCollectionService,
-                               Configuration hConf, StreamAdmin streamAdmin) {
+                               Configuration hConf, StreamAdmin streamAdmin, WorkflowToken workflowToken) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java]: Same as above.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java [hunk 6]
```diff
@@ -91,47 +96,55 @@ public String getExecutionMode() {
...
   public Transaction getTransaction() {
     return GSON.fromJson(hConf.get(HCONF_ATTR_NEW_TX), Transaction.class);
   }
 
+  /**
+   * @return the {@link WorkflowToken} stored in the configuration.
+   */
+  @Nullable
+  public WorkflowToken getWorkflowToken() {
+    return GSON.fromJson(hConf.get(HCONF_ATTR_WORKFLOW_TOKEN), BasicWorkflowToken.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java]: Will gson throws exception if `hConf.get()` returns null?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java [hunk 7]
```diff
@@ -155,4 +168,8 @@ private void setLogicalStartTime(long startTime) {
   private void setTransaction(Transaction tx) {
     hConf.set(HCONF_ATTR_NEW_TX, GSON.toJson(tx));
   }
+
+  public void setWorkflowToken(WorkflowToken workflowToken) {
+    hConf.set(HCONF_ATTR_WORKFLOW_TOKEN, GSON.toJson(workflowToken));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java]: Will there be exception if token is null? Also, the parameter should have `@Nullable`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java (hunk 6 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java]: No, tried a small program:
```
public static void main(String [] args) {
    Gson gson = new Gson();
    gson.toJson(null);
    System.out.println("hey");
    String json = null;
    gson.fromJson(json, BasicWorkflowToken.class);
    System.out.println("dingdong");
  }
```

No exceptions, ``GSON`` seems to handle it fine.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java (hunk 7 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextConfig.java]: No, Gson handles nulls. Added ``@Nullable``.

**Comment by chtyim** [comment]
LGTM

**Comment by bdmogal** [comment]
The last couple of commits won't impact unit test runtime. Squashing commits. Will wait for build - http://builds.cask.co/browse/CDAP-DUT2240-1 and merge.

**Comment by bdmogal** [comment]
Unrelated test (which passes locally) failure - http://builds.cask.co/browse/CDAP-DUT2240-JOB1-1. The build ran to the end as well. Reported https://issues.cask.co/browse/CDAP-3011.

Merging.

**Pull request merged**